sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```
@@ -0,0 +1,274 @@
+        for (i <- joinPlansLevelK.indices) {
+          val curJoinPlan = joinPlansLevelK(i)
+
+          val joinPlansOtherLevel = if (k == otherLevel) {
+            // Both sides of a join are at the same level, no need to repeat for previous ones.
+            joinPlansLevelK.drop(i)
+          } else {
+            foundPlans(otherLevel).values.toSeq
+          }
+
+          joinPlansOtherLevel.foreach { otherJoinPlan =>
+            // Should not join two overlapping item sets.
+            if (curJoinPlan.itemIds.intersect(otherJoinPlan.itemIds).isEmpty) {
+              val joinPlan = buildJoin(curJoinPlan, otherJoinPlan)
+              if (joinPlan.nonEmpty) {
+                // Check if it's the first plan for the item set, or it's a better plan than
+                // the existing one due to lower cost.
+                val existingPlan = foundPlansCurLevel.get(joinPlan.get.itemIds)
+                if (existingPlan.isEmpty || joinPlan.get.cost < existingPlan.get.cost) {
+                  foundPlansCurLevel.update(joinPlan.get.itemIds, joinPlan.get)
+                }
+              }
+            }
+          }
+        }
+
+        k += 1
+      }
+    }
+  }
+
+  /** Build a new join node. */
+  private def buildJoin(curJoinPlan: JoinPlan, otherJoinPlan: JoinPlan): Option[JoinPlan] = {
+    // Check if these two nodes are inner joinable. We consider cartesian product very
+    // costly, thus exclude such plans. This also helps us to reduce the search space.
+    val curPlan = curJoinPlan.plan
+    val otherPlan = otherJoinPlan.plan
+    val joinCond = conditions
+      .filterNot(l => canEvaluate(l, curPlan))
+      .filterNot(r => canEvaluate(r, otherPlan))
+      .filter(e => e.references.subsetOf(curPlan.outputSet ++ otherPlan.outputSet))
+
+    if (joinCond.nonEmpty) {
+      val curPlanStats = curPlan.stats(conf)
+      val otherPlanStats = otherPlan.stats(conf)
+      if (curPlanStats.rowCount.nonEmpty && otherPlanStats.rowCount.nonEmpty) {
```

Shouldn't try to fail early here? We should certainly check if we can fail early when we cannot do this for one of the inputs.